renderdoc/driver/gl/wrappers/gl_draw_funcs.cpp

```diff
@@ -709,10 +709,77 @@ bool WrappedOpenGL::Serialise_glDrawArrays(GLenum mode, GLint first, GLsizei cou
   return true;
 }
 
+std::function<void()> WrappedOpenGL::CopyClientMemoryArrays(GLint first, GLsizei count)
+{
+  ContextData &cd = GetCtxData();
+  GLResourceRecord *varecord = cd.m_VertexArrayRecord;
+  if(m_State != WRITING_CAPFRAME || varecord)    // Early out if VAO bound, as VAOs are VBO-only.
+    return []() {};
+
+  GLuint prevBinding;
+  m_Real.glGetIntegerv(eGL_ARRAY_BUFFER_BINDING, (GLint *)&prevBinding);
+
+  struct VertexAttribData
+  {
+    GLuint index;
+    GLint size;
+    GLenum type;
+    GLboolean normalized;
+    GLsizei stride;
+    void *pointer;
+  };
+  vector<VertexAttribData> attribs;
+  for(GLuint i = 0; i < ARRAY_COUNT(cd.m_ClientMemoryVBOs); i++)
+  {
+    GLint enabled = 0;
+    m_Real.glGetVertexAttribiv(i, eGL_VERTEX_ATTRIB_ARRAY_ENABLED, &enabled);
+    if(!enabled)
+      continue;
+
+    // Check that the attrib is using client-memory.
+    GLuint buffer;
+    m_Real.glGetVertexAttribiv(i, eGL_VERTEX_ATTRIB_ARRAY_BUFFER_BINDING, (GLint *)&buffer);
+    if(buffer != 0)
+      continue;
+
+    // App initially used client memory, so copy it into the temporary buffer.
+    VertexAttribData attrib;
+    memset(&attrib, 0, sizeof(attrib));
+    attrib.index = i;
+    m_Real.glGetVertexAttribiv(i, eGL_VERTEX_ATTRIB_ARRAY_SIZE, &attrib.size);
+    m_Real.glGetVertexAttribiv(i, eGL_VERTEX_ATTRIB_ARRAY_TYPE, (GLint *)&attrib.type);
+    m_Real.glGetVertexAttribiv(i, eGL_VERTEX_ATTRIB_ARRAY_NORMALIZED, (GLint *)&attrib.normalized);
+    m_Real.glGetVertexAttribiv(i, eGL_VERTEX_ATTRIB_ARRAY_STRIDE, &attrib.stride);
+    m_Real.glGetVertexAttribPointerv(i, eGL_VERTEX_ATTRIB_ARRAY_POINTER, &attrib.pointer);
+
+    GLint totalStride = attrib.stride ? attrib.stride : (GLint)GLTypeSize(attrib.type) * attrib.size;
+    glBindBuffer(eGL_ARRAY_BUFFER, cd.m_ClientMemoryVBOs[i]);
+    // Copy all client memory, and the pointer becomes a zero offset.
+    glBufferData(eGL_ARRAY_BUFFER, (first + count) * totalStride, attrib.pointer, eGL_STATIC_DRAW);
+    glVertexAttribPointer(attrib.index, attrib.size, attrib.type, attrib.normalized, attrib.stride,
+                          NULL);
+
+    attribs.push_back(attrib);
+  }
+
+  // Return a lambda to restore the 0-bindings and attrib pointers.
+  return [this, prevBinding, attribs]() {
+    glBindBuffer(eGL_ARRAY_BUFFER, 0);
+    for(const VertexAttribData &attrib : attribs)
+    {
+      glVertexAttribPointer(attrib.index, attrib.size, attrib.type, attrib.normalized,
+                            attrib.stride, attrib.pointer);
+    }
+    glBindBuffer(eGL_ARRAY_BUFFER, prevBinding);
+  };
+}
+
 void WrappedOpenGL::glDrawArrays(GLenum mode, GLint first, GLsizei count)
 {
   CoherentMapImplicitBarrier();
 
+  auto restore = CopyClientMemoryArrays(first, count);
```

Comment:
I went with the option of adding a RestoreClientMemoryArrays function.

Ive the left the struct name generic (ClientMemoryData) so that other kinds of arrays can be added later.